File: src/Form_ChatLog.vb
Imports System.Data.Common
Imports System.Data.SQLite
Imports System.IO
Imports System.Reflection
Imports System.Reflection.Emit
Imports System.Text
Imports System.Text.RegularExpressions
Imports WeifenLuo.WinFormsUI.Docking

Public Class Form_ChatLog
    Inherits DockContent
    Public Property MainFormInst As Form1
    Private Sub ChatLogForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        Dim mainForm As Form1 = CType(Application.OpenForms.OfType(Of Form1).FirstOrDefault(), Form1)
        If Not mainForm.cts.Token.IsCancellationRequested Then
            Me.Hide() ' フォームを閉じるのではなく、非表示にする
            e.Cancel = True ' イベントをキャンセルし、フォームが閉じないようにする
        Else
            e.Cancel = False
        End If
    End Sub
    Private Sub ChatLogForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'Form1.SetPictureBoxImage("TmCGPTD.iconChatBlack.png", PictureBox1)
        'Form1.SetPictureBoxImage("TmCGPTD.iconTagBlack.png", PictureBox2)

        ' 埋め込みリソースからボタン画像を読み込む
        Dim assembly As Assembly = Assembly.GetExecutingAssembly()
        Dim resourceStream As Stream = assembly.GetManifestResourceStream("TmCGPTD.iconCal.png")
        Dim originalImage As Image = Image.FromStream(resourceStream) ' 元の画像を読み込む
        Dim dpiScaling As Single = Form1.GetDpiScaling() ' DPIスケーリングを取得
        Dim newSize As Integer = CInt(26 * dpiScaling) ' 新しいサイズにリサイズする
        Dim resizedImage As New Bitmap(newSize, newSize)
        Using g As Graphics = Graphics.FromImage(resizedImage)
            g.InterpolationMode = Drawing2D.InterpolationMode.HighQualityBicubic
            g.DrawImage(originalImage, 0, 0, newSize - 2, newSize - 2)
        End Using
        ButtonCal.BackgroundImage = resizedImage ' ボタンのImageプロパティに設定
    End Sub

    ' データグリッド初期化--------------------------------------------------------------
    Private ReadOnly emptyDataTable As New DataTable()
    Public bSource As New BindingSource

    ' 仮想モードハンドラ
    Private Sub DataGrid_CellValueNeeded(sender As Object, e As DataGridViewCellValueEventArgs) Handles DataGrid.CellValueNeeded
        ' 必要なセルの値を取得して、e.Valueに設定する
        If e.ColumnIndex = 0 Then
            e.Value = e.RowIndex ' 行番号を設定
        Else
            e.Value = "Data" ' ダミーデータを設定
        End If
    End Sub
    Public Sub InitializeDataTable()

        ' 仮想モードを有効にする
        DataGrid.VirtualMode = True

        ' ダブルバッファリングを有効にする
        Dim dgvType As Type = DataGrid.GetType()
        Dim pi As PropertyInfo = dgvType.GetProperty("DoubleBuffered", BindingFlags.Instance Or BindingFlags.NonPublic)
        pi.SetValue(DataGrid, True, Nothing)

        emptyDataTable.Columns.Add("chatId", GetType(Integer))
        emptyDataTable.Columns.Add("chatDate", GetType(Date))
        emptyDataTable.Columns.Add("chatTitle", GetType(String))
        emptyDataTable.Columns.Add("chatTag", GetType(String))

        bSource.DataSource = emptyDataTable
        DataGrid.DataSource = bSource

        DataGrid.EnableHeadersVisualStyles = False
        'DataGrid.ColumnHeadersDefaultCellStyle.BackColor = Color.FromArgb(68, 70, 84)
        DataGrid.ColumnHeadersBorderStyle = DataGridViewHeaderBorderStyle.Single

        DataGrid.Columns(0).HeaderText = "ID"
        DataGrid.Columns(1).HeaderText = "Update"
        DataGrid.Columns(2).HeaderText = "Title"
        DataGrid.Columns(3).HeaderText = "Tags"

        DataGrid.Columns(1).DefaultCellStyle.Format = "yy/MM/dd HH:mm"

        DataGrid.GridColor = Color.FromArgb(84, 85, 99)

        Dim dpiScaleFactor As Single = CreateGraphics().DpiX / 96.0F

        DataGrid.Columns(0).Width = CInt(50 * dpiScaleFactor)
        DataGrid.Columns(1).Width = CInt(106 * dpiScaleFactor)
        DataGrid.Columns(2).Width = CInt(180 * dpiScaleFactor)
        DataGrid.Columns(3).Width = CInt(250 * dpiScaleFactor)
    End Sub

    ' 検索結果表示--------------------------------------------------------------
    Public Async Function ShowChatSearchResultAsync(dT As DataTable) As Task
        'If dT.Rows.Count > 0 Then
        Await Task.Run(Sub()
                           DataGrid.BeginInvoke(Sub()
                                                    ' 自動バインド停止
                                                    'bSource.RaiseListChangedEvents = False
                                                    bSource.SuspendBinding()

                                                    bSource.DataSource = dT ' データ入力

                                                    bSource.RaiseListChangedEvents = True ' データ変更許可
                                                    ' 更新反映
                                                    bSource.ResumeBinding()
                                                    'bSource.ResetBindings(False)
                                                    If DataGrid.Rows.Count > 0 Then DataGrid.Rows(0).Selected = True
                                                End Sub)
                       End Sub)
        'End If
        ' データベースからタグリストを取得
        Dim uniqueTags = Await MainFormInst.GetTagDatabaseAsync()
        Dim tagList As List(Of String) = uniqueTags.ToList()
        Await Task.Run(Sub() ComboBoxTag.BeginInvoke(Sub() ComboBoxTag.Items.Clear()))
        Await Task.Run(Sub() ComboBoxTag.BeginInvoke(Sub() ComboBoxTag.Items.AddRange(tagList.ToArray())))
    End Function

    ' チャット本文ログ表示--------------------------------------------------------------
    Private Async Sub DataGrid_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles DataGrid.CellClick
        If e.RowIndex >= 0 Then ' ヘッダー行を除く
            ' クリックされた行のchatIdカラムの値を取得
            Dim chatId As Integer = CInt(DataGrid.Rows(e.RowIndex).Cells("chatId").Value)
            Dim query As String = $"SELECT title, tag, json, text FROM chatlog WHERE id = {chatId}"
            Dim result As List(Of String) = Await MainFormInst.PickChatDatabaseAsync(query)
            Await MainFormInst.ShowChatLogAsync(result)
            MainFormInst.lastRowId = chatId
        End If
    End Sub

    ' データグリッドの自動選択解除--------------------------------------------------------------
    Private Sub Form_ChatLog_Activated(sender As Object, e As EventArgs) Handles MyBase.Activated
        If DataGrid.SelectedColumns.Count = 0 Then
            DataGrid.ClearSelection()
        End If
    End Sub

    Private keywordFilter As String = ""
    Private tagFilter As String = ""
    Private dateFilter As String = ""
    Public startDate As Date = Now
    Public endDate As Date = Now

    ' カレンダー起動--------------------------------------------------------------
    Private Async Sub ButtonCal_ClickAsync(sender As Object, e As EventArgs) Handles ButtonCal.Click
        Dim dateRangeForm As New Form_DateRange() ' DateRangeFormのインスタンスを作成
        Dim buttonScreenLocation As Point = ButtonCal.PointToScreen(Point.Empty) ' ボタンのスクリーン座標を取得
        ' ダイアログの表示位置を計算（ボタンの左下に右寄せ）
        Dim dialogLocation As New Point(buttonScreenLocation.X + CType(ButtonCal.Width, Integer) - CType(dateRangeForm.Width, Integer), buttonScreenLocation.Y + ButtonCal.Height)
        ' ダイアログの表示位置を設定
        dateRangeForm.Location = dialogLocation
        dateRangeForm.DateTimePicker1.Value = startDate
        dateRangeForm.DateTimePicker2.Value = endDate
        ' モーダルダイアログとして表示
        If dateRangeForm.ShowDialog() = DialogResult.OK Then
            ' OKボタンがクリックされた場合、選択された日付範囲を取得
            startDate = dateRangeForm.StartDatePicker.Value
            endDate = dateRangeForm.EndDatePicker.Value

            If endDate < startDate Then
                Dim tempDate As Date = startDate
                startDate = endDate
                endDate = tempDate
            End If
            Await ApplyDateFilterAsync(startDate, endDate)
            ButtonCal.BackColor = Color.FromArgb(146, 148, 165)
        Else
            Await RemoveDateFilterAsync() ' 日付フィルタ解除
            ButtonCal.BackColor = Color.FromArgb(53, 55, 64)
            startDate = Now
            endDate = Now
        End If
    End Sub
    ' 日付フィルタを適用するメソッド
    Private Async Function ApplyDateFilterAsync(startDate As Date, endDate As Date) As Task
        endDate = endDate.AddDays(1) ' endDate に1日加算
        dateFilter = $"chatDate >= #{startDate.ToShortDateString()}# AND chatDate <= #{endDate.ToShortDateString()}#"
        Await UpdateFilterAsync()
    End Function

    ' フィルタを解除するメソッド
    Private Async Function RemoveDateFilterAsync() As Task
        dateFilter = ""
        Await UpdateFilterAsync()
    End Function

    ' フィルタを更新するメソッド
    Private Async Function UpdateFilterAsync() As Task
        Dim filters As New List(Of String)

        If Not String.IsNullOrEmpty(keywordFilter) Then filters.Add(keywordFilter)
        If Not String.IsNullOrEmpty(tagFilter) Then filters.Add(tagFilter)
        If Not String.IsNullOrEmpty(dateFilter) Then filters.Add(dateFilter)

        Await Task.Run(Sub()
                           DataGrid.BeginInvoke(Sub()
                                                    ' 自動バインド停止
                                                    'bSource.RaiseListChangedEvents = False
                                                    bSource.SuspendBinding()

                                                    bSource.Filter = String.Join(" AND ", filters)

                                                    bSource.RaiseListChangedEvents = True ' データ変更許可
                                                    ' 更新反映
                                                    bSource.ResumeBinding()
                                                    'bSource.ResetBindings(False)
                                                    If DataGrid.Rows.Count > 0 Then DataGrid.Rows(0).Selected = True
                                                End Sub)
                       End Sub)
    End Function

    ' キーワード検索やタグ検索を更新する場合
    Private Async Sub ComboBoxTag_SelectedIndexChangedAync(sender As Object, e As EventArgs) Handles ComboBoxTag.SelectedIndexChanged
        ' 選択された項目を取得
        Dim selectedItem As String = Nothing
        Await Task.Run(Sub() ComboBoxTag.BeginInvoke(Sub() selectedItem = ComboBoxTag.SelectedItem?.ToString()))

        ' 空の場合は何もしない
        If String.IsNullOrEmpty(selectedItem) Then Return
        If selectedItem = "(All)" Then selectedItem = ""
        Await UpdateTagFilterAsync(selectedItem)
    End Sub
    Private Async Function UpdateKeywordFilterAsync(newKeyword As String) As Task
        keywordFilter = $"ColumnNameForKeyword LIKE '%{newKeyword}%'"
        Await UpdateFilterAsync()
    End Function

    Private Async Function UpdateTagFilterAsync(newTag As String) As Task
        tagFilter = $"chatTag LIKE '%{newTag}%'"
        Await UpdateFilterAsync()
    End Function

    Private Sub ComboBoxTag_DrawItem(sender As Object, e As DrawItemEventArgs) Handles ComboBoxTag.DrawItem
        If e.Index < 0 Then Return

        Dim comboBoxMod As ComboBox = CType(sender, ComboBox)
        Dim isSelected As Boolean = (e.State And DrawItemState.Selected) = DrawItemState.Selected

        ' カスタムの背景色とテキスト色を設定します。
        Dim backColor As Color = If(isSelected, Color.FromArgb(106, 108, 125), comboBoxMod.BackColor)
        Dim textColor As Color = If(isSelected, Color.FromArgb(220, 220, 220), comboBoxMod.ForeColor)

        ' 背景を描画します。
        Using brush As New SolidBrush(backColor)
            e.Graphics.FillRectangle(brush, e.Bounds)
        End Using

        ' テキストを描画します。
        Using brush As New SolidBrush(textColor)
            e.Graphics.DrawString(comboBoxMod.Items(e.Index).ToString(), e.Font, brush, e.Bounds)
        End Using

        ' フォーカス矩形を描画します。
        If (e.State And DrawItemState.Focus) = DrawItemState.Focus Then
            e.DrawFocusRectangle()
        End If
    End Sub

    ' インクリメンタルサーチ--------------------------------------------------------------
    Public Async Function TextBoxSearch_TextChangedAsync(sender As Object, e As EventArgs) As Task Handles TextBoxSearch.TextChanged
        Dim searchKey As String = Nothing
        Await Task.Run(Sub() TextBoxSearch.BeginInvoke(Sub() searchKey = TextBoxSearch.Text.Trim()))
        Dim query As String
        Dim dT As DataTable
        If searchKey = "" Then
            query = "SELECT id, date, title, tag FROM chatlog ORDER BY date DESC;"
            dT = Await MainFormInst.SearchChatDatabaseAsync(query)
        Else
            query = "SELECT id, date, title, tag FROM chatlog WHERE LOWER(text) LIKE LOWER(@searchKey)"
            dT = Await MainFormInst.SearchChatDatabaseAsync(query, searchKey)
        End If
        Await ShowChatSearchResultAsync(dT)
        Dim rowsCount As New Integer
        Await Task.Run(Sub() DataGrid.BeginInvoke(Sub() rowsCount = DataGrid.Rows.Count))
        If rowsCount > 0 Then
            Dim chatId As New Integer
            Await Task.Run(Sub() DataGrid.BeginInvoke(Sub() chatId = CInt(DataGrid.Rows(0).Cells("chatId").Value)))
            query = $"SELECT title, tag, json, text FROM chatlog WHERE id = {chatId}"
            Dim result = Await MainFormInst.PickChatDatabaseAsync(query)
            Await MainFormInst.ShowChatLogAsync(result, searchKey)
            MainFormInst.lastRowId = chatId
        Else

        End If
    End Function

    ' デリートボタン--------------------------------------------------------------
    Private Async Sub ButtonChatDelete_ClickAsync(sender As Object, e As EventArgs) Handles ButtonChatDelete.Click
        Dim chatId As Integer
        Dim chatDate As String
        Dim chatTitle As String
        If DataGrid.SelectedRows.Count > 0 Then
            Dim selectedRow As DataGridViewRow = DataGrid.SelectedRows(0)
            chatId = Convert.ToInt32(selectedRow.Cells("chatId").Value)
            chatDate = selectedRow.Cells("chatDate").Value.ToString
            chatTitle = selectedRow.Cells("chatTitle").Value.ToString
        Else
            Return
        End If

        Dim result = MessageBox.Show($"Delete this chat log?{Environment.NewLine}{Environment.NewLine}{chatDate} {chatTitle}", "Confirmation", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2)
        If result = DialogResult.Yes Then

            ' SQLクエリを構築して実行する
            Using connection As New SQLiteConnection($"Data Source={Form1.dbPath}")
                connection.Open()
                ' トランザクションを開始
                Using transaction As SQLiteTransaction = connection.BeginTransaction()
                    Try
                        ' レコードを削除
                        Using command As New SQLiteCommand($"DELETE FROM chatlog WHERE id = '{chatId}'", connection, transaction)
                            Await command.ExecuteNonQueryAsync()
                        End Using
                        ' トランザクションをコミット
                        transaction.Commit()
                    Catch ex As Exception
                        ' エラーが発生した場合は、トランザクションをロールバック
                        transaction.Rollback()
                        MsgBox(ex.Message)
                    End Try
                End Using
            End Using

            ' インメモリをいったん閉じてまた開く
            Form1.memoryConnection.Close()
            Form1.DbLoadToMemory()

            Await TextBoxSearch_TextChangedAsync(DataGrid, EventArgs.Empty)
        End If

    End Sub

    Private dockedSize As Size
    Private Sub MyForm_SizeChanged(sender As Object, e As EventArgs) Handles MyBase.SizeChanged
        If Me.DockState <> WeifenLuo.WinFormsUI.Docking.DockState.Float Then
            dockedSize = Me.Size
        End If
    End Sub
    Private Sub MyForm_DockStateChanged(sender As Object, e As EventArgs) Handles MyBase.DockStateChanged
        If Me.DockState = WeifenLuo.WinFormsUI.Docking.DockState.Float Then
            ' ここで DefaultFloatWindowSize を変更する
            Me.Size = dockedSize
        End If
    End Sub

End Class

File: src/Form_ChatLog.cs
using System.Data;
using System.Data.SQLite;
using System.Reflection;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using WeifenLuo.WinFormsUI.Docking;

namespace TmCGPTD
{

    public partial class Form_ChatLog : DockContent
    {
        public Form1 MainFormInst { get; set; }

        public Form_ChatLog()
        {
            InitializeComponent();
        }
        private void ChatLogForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            var mainForm = Application.OpenForms.OfType<Form1>().FirstOrDefault();
            if (!mainForm.cts.Token.IsCancellationRequested)
            {
                MainFormInst.chatLogForm.DockState = DockState.Hidden; // フォームを閉じるのではなく、非表示にする
                e.Cancel = true; // イベントをキャンセルし、フォームが閉じないようにする
            }
            else
            {
                e.Cancel = false;
            }
        }
        private void ChatLogForm_Load(object sender, EventArgs e)
        {
            // Form1.SetPictureBoxImage("TmCGPTD.iconChatBlack.png", PictureBox1)
            // Form1.SetPictureBoxImage("TmCGPTD.iconTagBlack.png", PictureBox2)

            // 埋め込みリソースからボタン画像を読み込む
            var assembly = Assembly.GetExecutingAssembly();
            var resourceStream = assembly.GetManifestResourceStream("TmCGPTD.images.iconCal.png");
            var originalImage = Image.FromStream(resourceStream); // 元の画像を読み込む
            float dpiScaling = MainFormInst.GetDpiScaling(); // DPIスケーリングを取得
            int newSize = (int)Math.Round(26f * dpiScaling); // 新しいサイズにリサイズする
            var resizedImage = new Bitmap(newSize, newSize);
            using (var g = Graphics.FromImage(resizedImage))
            {
                g.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
                g.DrawImage(originalImage, 0, 0, newSize - 2, newSize - 2);
            }
            ButtonCal.BackgroundImage = resizedImage; // ボタンのImageプロパティに設定
        }

        // データグリッド初期化--------------------------------------------------------------
        private readonly DataTable emptyDataTable = new DataTable();
        public BindingSource bSource = new BindingSource();

        // 仮想モードハンドラ
        private void DataGrid_CellValueNeeded(object sender, DataGridViewCellValueEventArgs e)
        {
            // 必要なセルの値を取得して、e.Valueに設定する
            if (e.ColumnIndex == 0)
            {
                e.Value = e.RowIndex; // 行番号を設定
            }
            else
            {
                e.Value = "Data";
            } // ダミーデータを設定
        }
        public void InitializeDataTable()
        {

            // 仮想モードを有効にする
            DataGrid.VirtualMode = true;

            // ダブルバッファリングを有効にする
            var dgvType = DataGrid.GetType();
            var pi = dgvType.GetProperty("DoubleBuffered", BindingFlags.Instance | BindingFlags.NonPublic);
            pi.SetValue(DataGrid, true, (object[])null);

            emptyDataTable.Columns.Add("chatId", typeof(int));
            emptyDataTable.Columns.Add("chatDate", typeof(DateTime));
            emptyDataTable.Columns.Add("chatTitle", typeof(string));
            emptyDataTable.Columns.Add("chatTag", typeof(string));

            bSource.DataSource = emptyDataTable;
            DataGrid.DataSource = bSource;

            DataGrid.EnableHeadersVisualStyles = false;
            // DataGrid.ColumnHeadersDefaultCellStyle.BackColor = Color.FromArgb(68, 70, 84)
            DataGrid.ColumnHeadersBorderStyle = DataGridViewHeaderBorderStyle.Single;

            DataGrid.Columns[0].HeaderText = "ID";
            DataGrid.Columns[1].HeaderText = "Update";
            DataGrid.Columns[2].HeaderText = "Title";
            DataGrid.Columns[3].HeaderText = "Tags";

            DataGrid.Columns[1].DefaultCellStyle.Format = "yy/MM/dd HH:mm";

            DataGrid.GridColor = Color.FromArgb(84, 85, 99);

            float dpiScaleFactor = CreateGraphics().DpiX / 96.0f;

            DataGrid.Columns[0].Width = (int)Math.Round(50f * dpiScaleFactor);
            DataGrid.Columns[1].Width = (int)Math.Round(106f * dpiScaleFactor);
            DataGrid.Columns[2].Width = (int)Math.Round(180f * dpiScaleFactor);
            DataGrid.Columns[3].Width = (int)Math.Round(250f * dpiScaleFactor);
        }

        // 検索結果表示--------------------------------------------------------------
        public async Task ShowChatSearchResultAsync(DataTable dT)
        {
            // If dT.Rows.Count > 0 Then
            await Task.Run(() => DataGrid.BeginInvoke(() =>
{
    // 自動バインド停止
    // bSource.RaiseListChangedEvents = False
    bSource.SuspendBinding();

    bSource.DataSource = dT; // データ入力

    bSource.RaiseListChangedEvents = true; // データ変更許可
                                           // 更新反映
    bSource.ResumeBinding();
    // bSource.ResetBindings(False)
    if (DataGrid.Rows.Count > 0)
        DataGrid.Rows[0].Selected = true;
}));
            // End If
            // データベースからタグリストを取得
            var uniqueTags = await MainFormInst.GetTagDatabaseAsync();
            var tagList = uniqueTags.ToList();
            await Task.Run(() => ComboBoxTag.BeginInvoke(() => ComboBoxTag.Items.Clear()));
            await Task.Run(() => ComboBoxTag.BeginInvoke(() => ComboBoxTag.Items.AddRange(tagList.ToArray())));
        }

        // チャット本文ログ表示--------------------------------------------------------------
        private async void DataGrid_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            if (e.RowIndex >= 0) // ヘッダー行を除く
            {
                // クリックされた行のchatIdカラムの値を取得
                int chatId = Conversions.ToInteger(DataGrid.Rows[e.RowIndex].Cells["chatId"].Value);
                string query = $"SELECT title, tag, json, text FROM chatlog WHERE id = {chatId}";
                var result = await MainFormInst.PickChatDatabaseAsync(query);
                await MainFormInst.ShowChatLogAsync(result);
                Form1.lastRowId = chatId;
            }
        }

        // データグリッドの自動選択解除--------------------------------------------------------------
        private void Form_ChatLog_Activated(object sender, EventArgs e)
        {
            if (DataGrid.SelectedColumns.Count == 0)
            {
                DataGrid.ClearSelection();
            }
        }

        private string keywordFilter = "";
        private string tagFilter = "";
        private string dateFilter = "";
        public DateTime startDate = DateTime.Now;
        public DateTime endDate = DateTime.Now;

        // カレンダー起動--------------------------------------------------------------
        private async void ButtonCal_ClickAsync(object sender, EventArgs e)
        {
            var dateRangeForm = new Form_DateRange(); // DateRangeFormのインスタンスを作成
            var buttonScreenLocation = ButtonCal.PointToScreen(Point.Empty); // ボタンのスクリーン座標を取得
                                                                             // ダイアログの表示位置を計算（ボタンの左下に右寄せ）
            var dialogLocation = new Point(buttonScreenLocation.X + ButtonCal.Width - dateRangeForm.Width, buttonScreenLocation.Y + ButtonCal.Height);
            // ダイアログの表示位置を設定
            dateRangeForm.Location = dialogLocation;
            dateRangeForm.DateTimePicker1.Value = startDate;
            dateRangeForm.DateTimePicker2.Value = endDate;
            // モーダルダイアログとして表示
            if (dateRangeForm.ShowDialog() == DialogResult.OK)
            {
                // OKボタンがクリックされた場合、選択された日付範囲を取得
                startDate = dateRangeForm.StartDatePicker.Value;
                endDate = dateRangeForm.EndDatePicker.Value;

                if (endDate < startDate)
                {
                    var tempDate = startDate;
                    startDate = endDate;
                    endDate = tempDate;
                }
                await ApplyDateFilterAsync(startDate, endDate);
                ButtonCal.BackColor = Color.FromArgb(146, 148, 165);
            }
            else
            {
                await RemoveDateFilterAsync(); // 日付フィルタ解除
                ButtonCal.BackColor = Color.FromArgb(53, 55, 64);
                startDate = DateTime.Now;
                endDate = DateTime.Now;
            }
        }
        // 日付フィルタを適用するメソッド
        private async Task ApplyDateFilterAsync(DateTime startDate, DateTime endDate)
        {
            endDate = endDate.AddDays(1d); // endDate に1日加算
            dateFilter = $"chatDate >= #{startDate.ToShortDateString()}# AND chatDate <= #{endDate.ToShortDateString()}#";
            await UpdateFilterAsync();
        }

        // フィルタを解除するメソッド
        private async Task RemoveDateFilterAsync()
        {
            dateFilter = "";
            await UpdateFilterAsync();
        }

        // フィルタを更新するメソッド
        private async Task UpdateFilterAsync()
        {
            var filters = new List<string>();

            if (!string.IsNullOrEmpty(keywordFilter))
                filters.Add(keywordFilter);
            if (!string.IsNullOrEmpty(tagFilter))
                filters.Add(tagFilter);
            if (!string.IsNullOrEmpty(dateFilter))
                filters.Add(dateFilter);

            await Task.Run(() => DataGrid.BeginInvoke(() =>
{
    // 自動バインド停止
    // bSource.RaiseListChangedEvents = False
    bSource.SuspendBinding();

    bSource.Filter = string.Join(" AND ", filters);

    bSource.RaiseListChangedEvents = true; // データ変更許可
                                           // 更新反映
    bSource.ResumeBinding();
    // bSource.ResetBindings(False)
    if (DataGrid.Rows.Count > 0)
        DataGrid.Rows[0].Selected = true;
}));
        }

        // キーワード検索やタグ検索を更新する場合
        private async void ComboBoxTag_SelectedIndexChangedAync(object sender, EventArgs e)
        {
            // 選択された項目を取得
            string selectedItem = null;
            await Task.Run(() => ComboBoxTag.BeginInvoke(() => selectedItem = ComboBoxTag.SelectedItem?.ToString()));

            // 空の場合は何もしない
            if (string.IsNullOrEmpty(selectedItem))
                return;
            if (selectedItem == "(All)")
                selectedItem = "";
            await UpdateTagFilterAsync(selectedItem);
        }
        private async Task UpdateKeywordFilterAsync(string newKeyword)
        {
            keywordFilter = $"ColumnNameForKeyword LIKE '%{newKeyword}%'";
            await UpdateFilterAsync();
        }

        private async Task UpdateTagFilterAsync(string newTag)
        {
            tagFilter = $"chatTag LIKE '%{newTag}%'";
            await UpdateFilterAsync();
        }

        private void ComboBoxTag_DrawItem(object sender, DrawItemEventArgs e)
        {
            if (e.Index < 0)
                return;

            ComboBox comboBoxMod = (ComboBox)sender;
            bool isSelected = (e.State & DrawItemState.Selected) == DrawItemState.Selected;

            // カスタムの背景色とテキスト色を設定します。
            var backColor = isSelected ? Color.FromArgb(106, 108, 125) : comboBoxMod.BackColor;
            var textColor = isSelected ? Color.FromArgb(220, 220, 220) : comboBoxMod.ForeColor;

            // 背景を描画します。
            using (var brush = new SolidBrush(backColor))
            {
                e.Graphics.FillRectangle(brush, e.Bounds);
            }

            // テキストを描画します。
            using (var brush = new SolidBrush(textColor))
            {
                e.Graphics.DrawString(comboBoxMod.Items[e.Index].ToString(), e.Font, brush, e.Bounds);
            }

            // フォーカス矩形を描画します。
            if ((e.State & DrawItemState.Focus) == DrawItemState.Focus)
            {
                e.DrawFocusRectangle();
            }
        }

        // インクリメンタルサーチ--------------------------------------------------------------
        public async void TextBoxSearch_TextChangedAsync(object sender, EventArgs e)
        {
            try
            {
                string searchKey = null;
                await Task.Run(() => TextBoxSearch.BeginInvoke(() => searchKey = TextBoxSearch.Text.Trim()));
                string query;
                DataTable dT;
                if (string.IsNullOrEmpty(searchKey))
                {
                    query = "SELECT id, date, title, tag FROM chatlog ORDER BY date DESC;";
                    dT = await MainFormInst.SearchChatDatabaseAsync(query);
                }
                else
                {
                    query = "SELECT id, date, title, tag FROM chatlog WHERE LOWER(text) LIKE LOWER(@searchKey)";
                    dT = await MainFormInst.SearchChatDatabaseAsync(query, searchKey);
                }
                await ShowChatSearchResultAsync(dT);
                int rowsCount = new int();
                await Task.Run(() => DataGrid.BeginInvoke(() => rowsCount = DataGrid.Rows.Count));
                if (rowsCount > 0)
                {
                    int chatId = new int();
                    await Task.Run(() => DataGrid.BeginInvoke(() => chatId = Conversions.ToInteger(DataGrid.Rows[0].Cells["chatId"].Value)));
                    query = $"SELECT title, tag, json, text FROM chatlog WHERE id = {chatId}";
                    var result = await MainFormInst.PickChatDatabaseAsync(query);
                    await MainFormInst.ShowChatLogAsync(result, searchKey);
                    Form1.lastRowId = chatId;
                }
                else
                {

                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox( "Error: " + ex.Message, MsgBoxStyle.Exclamation, "Error");
            }
        }

        // デリートボタン--------------------------------------------------------------
        private async void ButtonChatDelete_ClickAsync(object sender, EventArgs e)
        {
            int chatId;
            string chatDate;
            string chatTitle;
            if (DataGrid.SelectedRows.Count > 0)
            {
                var selectedRow = DataGrid.SelectedRows[0];
                chatId = Convert.ToInt32(selectedRow.Cells["chatId"].Value);
                chatDate = selectedRow.Cells["chatDate"].Value.ToString();
                chatTitle = selectedRow.Cells["chatTitle"].Value.ToString();
            }
            else
            {
                return;
            }

            var result = MessageBox.Show($"Delete this chat log?{Environment.NewLine}{Environment.NewLine}{chatDate} {chatTitle}", "Confirmation", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2);
            if (result == DialogResult.Yes)
            {

                // SQLクエリを構築して実行する
                using (var connection = new SQLiteConnection($"Data Source={Form1.dbPath}"))
                {
                    connection.Open();
                    // トランザクションを開始
                    using (var transaction = connection.BeginTransaction())
                    {
                        try
                        {
                            // レコードを削除
                            using (var command = new SQLiteCommand($"DELETE FROM chatlog WHERE id = '{chatId}'", connection, transaction))
                            {
                                await command.ExecuteNonQueryAsync();
                            }
                            // トランザクションをコミット
                            transaction.Commit();
                        }
                        catch (Exception ex)
                        {
                            // エラーが発生した場合は、トランザクションをロールバック
                            transaction.Rollback();
                            Interaction.MsgBox(ex.Message);
                        }
                    }
                }

                // インメモリをいったん閉じてまた開く
                Form1.memoryConnection.Close();
                MainFormInst.DbLoadToMemory();

                TextBoxSearch_TextChangedAsync(DataGrid, EventArgs.Empty);
            }

        }

        private Size dockedSize;
        private void MyForm_SizeChanged(object sender, EventArgs e)
        {
            if (DockState != DockState.Float)
            {
                dockedSize = Size;
            }
        }
        private void MyForm_DockStateChanged(object sender, EventArgs e)
        {
            if (DockState == DockState.Float)
            {
                // ここで DefaultFloatWindowSize を変更する
                Size = dockedSize;
            }
        }

    }
}